pkg/database/diagnosticNode.go

```diff
@@ -0,0 +1,69 @@
+package database
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+
+	"github.com/QMSTR/go-qmstr/service"
+)
+
+// AddDiagnosticNodes stores the given DiagnosticNodes in a PackageNode or FileNode identified by the nodeID
+func (db *DataBase) AddDiagnosticNodes(nodeID string, diagnosticnodes ...*service.DiagnosticNode) error {
+	db.insertMutex.Lock()
+	defer db.insertMutex.Unlock()
+
+	const q = `
+	query Node($id: string){
+		node(func: uid($id)) @filter(has(packageNodeType) or has(fileNodeType)) @recurse(loop: false) {
+			uid
+			diagnosticInfo
+			packageNodeType
+			fileNodeType
+		}
+	}
+	`
+	vars := map[string]string{"$id": nodeID}
+	var result map[string][]interface{}
+
+	resp, err := db.client.NewTxn().QueryWithVars(context.Background(), q, vars)
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(resp.Json, &result)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if len(result["node"]) < 1 {
+		return fmt.Errorf("No package or file node with uid %s found", nodeID)
+	}
+
+	receiverNode := result["node"][0].(map[string]interface{})
+	var diagnosticInfo []*service.DiagnosticNode
+	if diagnosticInfoInter, ok := receiverNode["diagnosticInfo"]; ok {
+		diagnosticInfo = diagnosticInfoInter.([]*service.DiagnosticNode)
+	}
+	diagnosticInfo = append(diagnosticInfo, diagnosticnodes...)
+
+	if _, ok := receiverNode["packageNodeType"]; ok {
+		packageNode := service.PackageNode{}
+		packageNode.Uid = nodeID
+		packageNode.DiagnosticInfo = diagnosticInfo
+		_, err = dbInsert(db.client, &packageNode)
+		if err != nil {
+			return err
+		}
+	} else {
+		fileNode := service.FileNode{}
```

Comment:
Isn't this meaningless to create a node to attach a diagnostic node to it?

Reply:
The node has the uid of the existing file node, so we don't create a new filenode, we just attach the new diagnostic nodes. That's how we also add the infonodes in the db.

Reply:
I see. But the else block is misleading -- like all almost all else statements. We should instead have another if to check for fileNodeType -- since you actually querying for it -- and in the end produce an error if the node was neither a packagenode nor a filenode. This way it will become clear and extendable. We might want to add diagnostic information to analyzer nodes at some point e.g. deprecation warnings etc.

Reply:
@hemarkus Done. Btw I already connect diagnostic nodes to analyzer nodes.